Question: How to setup the client to access cluster wide (custom) resources

I added a controller to our Kubernetes cluster which operates at the cluster wide level. I want to create object with this client.
So I created java code from the controller's crd's.

When i try to create an object, the client seems to be namespace-scoped by default.
For example:
https://xxxx:xxxx/apis/capsule.clastix.io/v1beta1/namespaces/default/tenants

This returns a 404 because this path does not exist at the api-server.

I expect it to be:
https://xxxx:xxxx/apis/capsule.clastix.io/v1beta1/tenants

How can I instrument the client so it won't use the namespace while creating the url.

```java
        KubernetesClient client = new KubernetesClientBuilder().build();

        MixedOperation<io.clastix.capsule.v1beta1.Tenant,
                KubernetesResourceList<io.clastix.capsule.v1beta1.Tenant>, Resource<io.clastix.capsule.v1beta1.Tenant>>
                tenantClient = client.resources(io.clastix.capsule.v1beta1.Tenant.class);
       
       .....
         
       Object bla = tenantClient.resource(t).create();
```

I noticed while creating the java classes with the `java-generator-cli-6.4.0.sh` the Tenant class implements the `io.fabric8.kubernetes.api.model.Namespaced` interface. If I remove the interface, things work as expected..

Thx

Hi @MartijnStraatman , sorry for the delay, I missed this issue.
With a script like the:

```bash
#! /bin/bash
set -x

CRD=$1
CRD_NAME="${2:-default}"

# .XXXXXXX makes it cross-compatible Mac/Linux
TMP_DIR=$(mktemp -d -t $CRD_NAME.XXXXXXX)

cp ${CRD} $TMP_DIR/$CRD_NAME.yaml
mkdir -p $TMP_DIR/src

jbang io.fabric8:java-generator-cli:6.4.1 -u https://raw.githubusercontent.com/clastix/capsule/master/charts/capsule/crds/tenant-crd.yaml --target=$TMP_DIR/src

cat <<EOF >> $TMP_DIR/$CRD_NAME.java
//DEPS io.fabric8:kubernetes-client:6.2.0
//DEPS io.fabric8:generator-annotations:6.2.0
//DEPS io.sundr:builder-annotations:0.90.4
//DEPS org.projectlombok:lombok:1.18.24
//DEPS io.sundr:builder-annotations:0.90.4
//DEPS javax.validation:validation-api:2.0.1.Final
//SOURCES src/**.java
EOF

(
  cd $TMP_DIR
  jbang export portable --force $TMP_DIR/$CRD_NAME.java
  jbang --class-path $TMP_DIR/$CRD_NAME.jar -i
)

rm -rf $TMP_DIR
```

you have access to a shell session where you can reproduce the issue:

```java
import io.fabric8.kubernetes.client.*
var client = new KubernetesClientBuilder().build();
client.resources(io.clastix.capsule.v1beta1.Tenant.class).create(new io.clastix.capsule.v1beta1.Tenant())
```

You get the following StackTrace:
```
|  Exception io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: POST at: https://127.0.0.1:49635/apis/capsule.clastix.io/v1beta1/namespaces/default/tenants. Message: Not Found.
|        at KubernetesClientException.copyAsCause (KubernetesClientException.java:238)
|        at OperationSupport.waitForResult (OperationSupport.java:517)
|        at OperationSupport.handleResponse (OperationSupport.java:551)
|        at OperationSupport.handleResponse (OperationSupport.java:535)
|        at OperationSupport.handleCreate (OperationSupport.java:328)
|        at BaseOperation.handleCreate (BaseOperation.java:675)
|        at BaseOperation.handleCreate (BaseOperation.java:88)
|        at CreateOnlyResourceOperation.create (CreateOnlyResourceOperation.java:42)
|        at (#3:1)
|  Caused by: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: POST at: https://127.0.0.1:49635/apis/capsule.clastix.io/v1beta1/namespaces/default/tenants. Message: Not Found.
|        at OperationSupport.requestFailure (OperationSupport.java:709)
|        at OperationSupport.requestFailure (OperationSupport.java:689)
|        at OperationSupport.assertResponseCode (OperationSupport.java:640)
|        at OperationSupport.lambda$handleResponse$0 (OperationSupport.java:576)
|        at CompletableFuture$UniApply.tryFire (CompletableFuture.java:642)
|        at CompletableFuture.postComplete (CompletableFuture.java:506)
|        at CompletableFuture.complete (CompletableFuture.java:2073)
|        at OperationSupport.lambda$retryWithExponentialBackoff$2 (OperationSupport.java:618)
|        at CompletableFuture.uniWhenComplete (CompletableFuture.java:859)
|        at CompletableFuture$UniWhenComplete.tryFire (CompletableFuture.java:837)
|        at CompletableFuture.postComplete (CompletableFuture.java:506)
|        at CompletableFuture.complete (CompletableFuture.java:2073)
|        at OkHttpClientImpl$4.onResponse (OkHttpClientImpl.java:277)
|        at RealCall$AsyncCall.execute (RealCall.java:203)
|        at NamedRunnable.run (NamedRunnable.java:32)
|        at ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1128)
|        at ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:628)
|        at Thread.run (Thread.java:829)
```

cc. @manusa @shawkins

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,7 @@
 #### Bugs
 
 #### Improvements
+* Fix #4800: [java-generator] Reflect the `scope` field when implementing the `Namespaced` interface
 * Fix #4739: honor optimistic concurrency control semantics in the mock server for `PUT` and `PATCH` requests.
 * Fix #4644: generate CRDs in parallel and optimize code
 * Fix #4795: don't print warning message when service account token property is unset

--- a/java-generator/core/src/main/java/io/fabric8/java/generator/CRGeneratorRunner.java
+++ b/java-generator/core/src/main/java/io/fabric8/java/generator/CRGeneratorRunner.java
@@ -44,6 +44,7 @@ public class CRGeneratorRunner {
     CustomResourceDefinitionSpec crSpec = crd.getSpec();
     String crName = crSpec.getNames().getKind();
     String group = crSpec.getGroup();
+    String scope = crSpec.getScope();
 
     List<WritableCRCompilationUnit> writableCUs = new ArrayList<>(crSpec.getVersions().size());
     for (CustomResourceDefinitionVersion crdv : crSpec.getVersions()) {
@@ -84,6 +85,7 @@ public class CRGeneratorRunner {
           crName,
           group,
           version,
+          scope,
           prefix + "Spec",
           prefix + "Status",
           specGenerator != null,

--- a/java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JCRObject.java
+++ b/java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JCRObject.java
@@ -33,6 +33,7 @@ public class JCRObject extends AbstractJSONSchema2Pojo implements JObjectExtraAn
   private final String className;
   private final String group;
   private final String version;
+  private final String scope;
   private final String specClassName;
   private final String statusClassName;
   private final boolean withSpec;
@@ -48,6 +49,7 @@ public class JCRObject extends AbstractJSONSchema2Pojo implements JObjectExtraAn
       String type,
       String group,
       String version,
+      String scope,
       String specClassName,
       String statusClassName,
       boolean withSpec,
@@ -64,6 +66,7 @@ public class JCRObject extends AbstractJSONSchema2Pojo implements JObjectExtraAn
     this.className = type;
     this.group = group;
     this.version = version;
+    this.scope = scope;
     this.specClassName = specClassName;
     this.statusClassName = statusClassName;
     this.withSpec = withSpec;
@@ -131,7 +134,10 @@ public class JCRObject extends AbstractJSONSchema2Pojo implements JObjectExtraAn
         .setTypeArguments(spec, status);
 
     clz.addExtendedType(crType);
-    clz.addImplementedType("io.fabric8.kubernetes.api.model.Namespaced");
+
+    if (scope.equals("Namespaced")) {
+      clz.addImplementedType("io.fabric8.kubernetes.api.model.Namespaced");
+    }
 
     if (config.isGeneratedAnnotations()) {
       clz.addAnnotation(GENERATED_ANNOTATION);

--- a/java-generator/core/src/test/java/io/fabric8/java/generator/GeneratorTest.java
+++ b/java-generator/core/src/test/java/io/fabric8/java/generator/GeneratorTest.java
@@ -61,6 +61,7 @@ class GeneratorTest {
         "t",
         "g",
         "v",
+        "Namespaced",
         "Spec",
         "Status",
         true,
@@ -82,6 +83,59 @@ class GeneratorTest {
   }
 
   @Test
+  void testNamespacedCR() {
+    // Arrange
+    JCRObject cro = new JCRObject(
+        null,
+        "t",
+        "g",
+        "v",
+        "Namespaced",
+        "Spec",
+        "Status",
+        true,
+        true,
+        true,
+        true,
+        "t",
+        "ts",
+        defaultConfig);
+
+    // Act
+    GeneratorResult res = cro.generateJava();
+
+    // Assert
+    assertEquals("io.fabric8.kubernetes.api.model.Namespaced", res.getTopLevelClasses().get(0).getCompilationUnit()
+        .getClassByName("t").get().getImplementedTypes().get(0).getNameWithScope());
+  }
+
+  @Test
+  void testClusterScopeCR() {
+    // Arrange
+    JCRObject cro = new JCRObject(
+        null,
+        "t",
+        "g",
+        "v",
+        "Cluster",
+        "Spec",
+        "Status",
+        true,
+        true,
+        true,
+        true,
+        "t",
+        "ts",
+        defaultConfig);
+
+    // Act
+    GeneratorResult res = cro.generateJava();
+
+    // Assert
+    assertTrue(res.getTopLevelClasses().get(0).getCompilationUnit().getClassByName("t").get().getImplementedTypes().isEmpty());
+  }
+
+  @Test
   void testCRWithoutNamespace() {
     // Arrange
     JCRObject cro = new JCRObject(
@@ -89,6 +143,7 @@ class GeneratorTest {
         "t",
         "g",
         "v",
+        "Namespaced",
         "Spec",
         "Status",
         true,

